utils/test_bootstrapping/test_bootstrap_keygen.cpp

```diff
@@ -0,0 +1,175 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+#include <helib/helib.h>
+#include <helib/debugging.h>
+
+#include <helib/ArgMap.h>
+
+#include <random>
+
+#include <common.h>
+
+helib::Ptxt<helib::BGV> generateRandomPtxt(const helib::Context& context)
+{
+  // Note: this function uses a single coefficient in each slot.  For this
+  // reason there is no difference between THIN and THICK bootstrapping.
+  // Consider generating random polynomials if THICK
+  helib::Ptxt<helib::BGV> ptxt(context);
+  std::mt19937 gen(231087);
+  std::uniform_int_distribution<int> coinFlipDist(0, context.getP() - 1);
+  for (std::size_t i = 0; i < ptxt.size(); ++i) {
+    ptxt[i] = coinFlipDist(gen);
+  }
+  return ptxt;
+}
+
+int main(int argc, char** argv)
+{
+  std::string sk_file_name;
+  std::string pk_file_name;
+
+  bool thick = false;
+  bool quiet = false;
+  bool sk_only = false;
+
+  // clang-format off
+  helib::ArgMap()
+        .toggle()
+          .arg("--thick", thick, "perform thick bootstrapping", nullptr)
+          .arg("--quiet", quiet, "Suppress most of the output", nullptr)
+          .arg("-s", sk_only, "only the secret key polynomial is written to the secret key file.")
+        .positional()
+        .required()
+          .arg("<sk-file>", sk_file_name,
+               "The file containing context and secret key.", nullptr)
+          .arg("<pk-file>", pk_file_name, "The file containing context and evaluation public key.", nullptr)
+        .parse(argc, argv);
+  // clang-format on
+
+  
+  std::unique_ptr<helib::Context> pk_contextp;
+  std::unique_ptr<helib::PubKey> pkp;
+  
+  try {
+    std::tie(pk_contextp, pkp) = loadContextAndKey<helib::PubKey>(pk_file_name);
+  } catch (const helib::RuntimeError& e) {
+    std::cerr << e.what() << std::endl;
+    return EXIT_FAILURE;
+  }
+  
+  // Load Context and PubKey
+  helib::Context& context(*pk_contextp);
+  helib::PubKey& publicKey(*pkp);
+
+  if (!context.isBootstrappable()) {
+    std::cerr << "Context is not bootstrappable" << std::endl;
+    return EXIT_FAILURE;
+  }
+  const helib::EncryptedArray& ea(context.getEA());
+  helib::Ptxt<helib::BGV> ptxt =
+      generateRandomPtxt(context); // Random in [0, p)
+  helib::Ctxt ctxt(publicKey);
+  publicKey.Encrypt(ctxt, ptxt);
+
+  // Start doing bootstrapping
+  if (!quiet) {
+    std::cout << "Initial bootstrap" << std::endl;
+  }
+  if (thick) {
+    publicKey.reCrypt(ctxt);
+  } else {
+    publicKey.thinReCrypt(ctxt);
+  }
+  if (!quiet) {
+    helib::CheckCtxt(ctxt, "After recryption");
+    std::cout << "ctxt.bitCapacity: " << ctxt.bitCapacity() << std::endl;
+  }
+
+  int round = 0;
```

Comment:
duplicate?

Reply:
Yes test_bootstrap_keygen is a duplicate of test_bootstrap with the exception that it has separate inputs for evaluation keys and decryption key. Is that what you mean?